# Khám phá API vnstock — Ngày 2

Mục tiêu: hiểu dữ liệu `vnstock` trả về trước khi viết pipeline.
Không viết code chính thức ở đây.

Ngày chạy: 2026-08-31 · Thư viện: vnstock 4.0.7


In [1]:
from vnstock import Vnstock

## 1. Lấy thử 1 mã, 6 tháng

Gọi `Quote(symbol='VNM', source='VCI').history(...)`.

- symbol='VNM' — mã cổ phiếu Vinamilk.
- source='VCI' — nguồn dữ liệu.
- interval='1D' — nến ngày.

In [2]:
stock = Vnstock().stock(symbol='VNM', source='VCI')
df = stock.quote.history(start='2024-01-01', end='2024-06-30', interval='1D')



  ╭──────────────────────────────────────────────────────────╮
  │  ⚠️  VNSTOCK DEPRECATION NOTICE (31/08/2025)             │
  │                                                          │
  │  Lớp Vnstock và các phương thức cũ (stock, fx, crypto,   │
  │  world_index, fund...) đã chính thức bị ngừng hỗ trợ.    │
  │                                                          │
  │  Để hệ thống ổn định và nhận được cập nhật mới nhất,     │
  │  vui lòng chuyển sang dùng bộ thư viện `vnstock.api`.    │
  │                                                          │
  │  👉 Xem hướng dẫn Migration: /vnstock-migration          │
  ╰──────────────────────────────────────────────────────────╯

Mẫu code chuyển đổi (Migration Example):
--------------------------------------
Cũ (Old):  stock = Vnstock().stock('ACB')
Mới (New): from vnstock.api.quote import Quote
          q = Quote(symbol='ACB', source='VCI')



In [3]:
df

,time,open,high,low,close,volume
0,2023-12-14,57.41,57.83,57.24,57.41,1463160
1,2023-12-15,57.49,57.58,57.16,57.16,3741675
2,2023-12-18,57.16,57.24,56.23,56.23,3518371
3,2023-12-19,56.32,56.57,55.65,56.23,3066185
4,2023-12-20,56.23,57.24,56.23,57.16,2605917
...,...,...,...,...,...,...
127,2024-06-24,56.62,57.05,55.93,56.10,3851919
128,2024-06-25,56.36,56.70,56.10,56.10,2436399
129,2024-06-26,56.36,56.36,55.76,55.85,2488180
130,2024-06-27,55.85,56.19,55.76,55.93,2710114


## 2. Lớp Vnstock() cũ đã bị khi tử từ  31/08/2025. Chuyển sang API mới vnstock.api.quote và thử lại.
Cũ (Old):  stock = Vnstock().stock('ACB')
Mới (New): from vnstock.api.quote import Quote
          q = Quote(symbol='ACB', source='VCI')

In [3]:
# Thêm vào để hiện kết quả toàn bộ dòng trong cell
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"


In [4]:
from vnstock.api.quote import Quote

In [5]:
q = Quote(symbol='VNM', source='VCI' ) 
df = q.history(start='2024-01-01', end='2024-06-30', interval='1D')
df

,time,open,high,low,close,volume
0,2023-12-14,57.41,57.83,57.24,57.41,1463160
1,2023-12-15,57.49,57.58,57.16,57.16,3741675
2,2023-12-18,57.16,57.24,56.23,56.23,3518371
3,2023-12-19,56.32,56.57,55.65,56.23,3066185
4,2023-12-20,56.23,57.24,56.23,57.16,2605917
...,...,...,...,...,...,...
127,2024-06-24,56.62,57.05,55.93,56.10,3851919
128,2024-06-25,56.36,56.70,56.10,56.10,2436399
129,2024-06-26,56.36,56.36,55.76,55.85,2488180
130,2024-06-27,55.85,56.19,55.76,55.93,2710114


### Đã chạy được API mới và lấy được dữ liệu.

In [ ]:
print(type(df))     # Kiểu trả về: <class 'pandas.core.frame.DataFrame'>
df.shape            # 132 dòng, 6 cột
df.columns          # Tên cột chính xác: `time`, `open`, `high`, `low`, `close`, `volume`
df.dtypes           # kiểu mỗi cột: time - datetime64[ns]; open, high, low, close - float64; volume - int64; dtype: object
df.head(10)         # 10 dòng đầu
df.tail(10)         # 10 dòng cuối
df.isnull().sum()   # mỗi cột thiếu bao nhiêu giá trị
df.describe()       # min/max/mean: các giá trị hợp lý
df['time'].min(), df['time'].max()   #  khoảng ngày: Timestamp('2023-12-14 00:00:00'), Timestamp('2024-06-28 00:00:00')

<class 'pandas.core.frame.DataFrame'>


(132, 6)

Index(['time', 'open', 'high', 'low', 'close', 'volume'], dtype='object')

time      datetime64[ns]
open             float64
high             float64
low              float64
close            float64
volume             int64
dtype: object

,time,open,high,low,close,volume
0,2023-12-14,57.41,57.83,57.24,57.41,1463160
1,2023-12-15,57.49,57.58,57.16,57.16,3741675
2,2023-12-18,57.16,57.24,56.23,56.23,3518371
3,2023-12-19,56.32,56.57,55.65,56.23,3066185
4,2023-12-20,56.23,57.24,56.23,57.16,2605917
5,2023-12-21,56.99,57.33,56.91,56.91,2587654
6,2023-12-22,56.91,57.16,56.65,56.74,2147607
7,2023-12-25,56.74,57.66,56.74,56.99,2675110
8,2023-12-26,57.33,57.49,57.07,57.07,1977559
9,2023-12-27,57.50,57.67,57.08,57.08,1628408


,time,open,high,low,close,volume
122,2024-06-17,56.79,57.05,56.53,56.53,3621867
123,2024-06-18,56.96,56.96,56.45,56.45,2754686
124,2024-06-19,56.53,56.53,55.76,55.93,6160999
125,2024-06-20,56.02,57.22,55.85,56.53,4400155
126,2024-06-21,56.96,57.47,56.53,56.53,3970969
127,2024-06-24,56.62,57.05,55.93,56.10,3851919
128,2024-06-25,56.36,56.70,56.10,56.10,2436399
129,2024-06-26,56.36,56.36,55.76,55.85,2488180
130,2024-06-27,55.85,56.19,55.76,55.93,2710114
131,2024-06-28,56.10,56.53,55.93,56.10,2735071


time      0
open      0
high      0
low       0
close     0
volume    0
dtype: int64

,time,open,high,low,close,volume
count,132,132.000000,132.000000,132.000000,132.000000,1.320000e+02
mean,2024-03-23 03:05:27.272727296,57.628106,58.042727,57.199091,57.493409,3.686726e+06
min,2023-12-14 00:00:00,54.820000,55.080000,53.960000,54.390000,1.230002e+06
25%,2024-01-30 18:00:00,56.777500,57.130000,56.450000,56.620000,2.458189e+06
50%,2024-03-23 12:00:00,57.315000,57.665000,56.990000,57.160000,3.261988e+06
75%,2024-05-14 06:00:00,58.240000,58.592500,57.655000,58.070000,4.486209e+06
max,2024-06-28 00:00:00,61.980000,62.150000,60.710000,61.730000,1.187643e+07
std,NaN,1.476670,1.529294,1.421243,1.470865,1.754797e+06


(Timestamp('2023-12-14 00:00:00'), Timestamp('2024-06-28 00:00:00'))

## 3. Các trường hợp biên

In [19]:
# Mã không tồn tại
Quote(symbol='XXXXX', source='VCI').history(start='2024-01-01', end='2024-06-30')


ValueError: Invalid symbol. Your symbol format is not recognized!

In [25]:
Quote(symbol='ZZZ', source='VCI').history(start='2024-01-01', end='2024-06-30')

RetryError: RetryError[<Future at 0x1d561f0f990 state=finished raised ValueError>]

In [6]:
# Chỉ ngày nghỉ Tết
Quote(symbol='VNM', source='VCI').history(start='2024-02-09', end='2024-02-12', interval='1D')


,time,open,high,low,close,volume
0,2024-02-02,57.08,57.25,56.65,56.99,2613144
1,2024-02-05,56.99,57.67,56.82,57.50,3568444
2,2024-02-06,57.58,57.67,57.16,57.16,2394933
3,2024-02-07,57.16,57.41,56.82,57.16,2737292


In [8]:
# Ngày tương lai
d = Quote(symbol='VNM', source='VCI').history(start='2030-01-01', end='2030-01-10', interval='1D')
d['time']

0   2026-08-17
1   2026-08-18
2   2026-08-19
3   2026-08-20
4   2026-08-21
5   2026-08-24
6   2026-08-25
7   2026-08-26
8   2026-08-27
9   2026-08-28
Name: time, dtype: datetime64[ns]

In [10]:
d = Quote(symbol='VNM', source='VCI').history(start='2024-02-10', end='2024-02-12', interval='1D')
d.empty        # True hay False?
d['time']      # nếu không rỗng thì ra ngày nào?


False

0   2024-02-05
1   2024-02-06
2   2024-02-07
Name: time, dtype: datetime64[ns]

### Hành vi khi lỗi

| Trường hợp        | Kết quả                                                            |
|-------------------|--------------------------------------------------------------------|
| Mã sai định dạng (XXXXX) | Ném `ValueError: Invalid symbol...` ngay khi tạo `Quote()`  |
| Mã đúng định dạng, không tồn tại (ZZZ) | Ném `RetryError: RetryError...`               |
| Khoảng toàn ngày nghỉ Tết | Không rỗng, trả về `ngày giao dịch gần đó` (sàn vẫn nghỉ)  |
| Khoảng ngày tương lai | Trả về `10 phiên gần nhất`, kết thúc ở `phiên mới nhất có thật`|

-> API không tự báo lỗi khi khoảng ngày không có phiên giao dịch. Nó âm thầm thay bằng các phiên gần nhất (trong hoặc trước khoảng yêu cầu). Số dòng trả về xấp xỉ số ngày trong khoảng yêu cầu. Pipeline phải tự lọc df theo [start, end] và validator phải chặn ngày tương lai / ngoài khoảng. Chỉ mã sai định dạng mới ném lỗi → cần try/except ValueError khi gọi.

API chậm và có rate limit khi viết lại `d = Quote...`, `d['time']`  pipeline phải (1) lưu dữ liệu thô xuống đĩa để không gọi lại, (2) có retry + chờ tăng dần khi bị chặn. Thoát ra chạy lại từ dầu thì không bị chậm nữa. 

### Suy ra cho pipeline
- Không gọi API dồn dập. Cần nghỉ giữa các lần gọi (rate limiting chủ động).
- Cần retry + chờ tăng dần (exponential backoff), giới hạn số lần thử.
- Cần lưu dữ liệu thô xuống đĩa để không phải gọi lại khi chạy lại transform.

- Nguồn VCI không cắt chặt theo `start`: xin từ 2024-01-01 nhưng trả về từ 2023-11-13 (~1.5 tháng thừa ở đầu). `end` cho đúng ngày nhưng là 0h.
- Kết hợp với các phát hiện trước: KHÔNG được tin biên ngày mà API trả về.


## 4. So sánh nguồn dữ liệu: VCI vs kbs

### Vì sao phải đổi nguồn

Khi viết `main.py` (Ngày 3), xin dữ liệu từ `2024-01-01` với `source='VCI'`
nhưng nhận về bảng bắt đầu từ `2023-11-13` — thừa ~1.5 tháng ở đầu khoảng.
`end` thì được tôn trọng, chỉ `start` bị bỏ qua.

Thử đổi sang `source='kbs'`: xin từ `2026-01-01`, nhận về từ `2026-01-05`
— đúng phiên giao dịch đầu tiên sau ngày yêu cầu (01/01 nghỉ Tết dương,
03–04/01 là cuối tuần).

**→ Chọn `kbs` làm nguồn chính vì cắt đúng biên ngày yêu cầu.**

| Hành vi | VCI | kbs |
|---|---|---|
| Tôn trọng `start` | KHÔNG (trả thừa ~1.5 tháng đầu) | CÓ |
| Tôn trọng `end` | Có | Có |
| Kiểu cột `time` | `datetime64` — date thuần (2023-11-13) | `datetime64` — CÓ giờ (2026-01-05 07:00:00) |

### Vấn đề mới của kbs: cột `time` có phần giờ `07:00:00`

Toàn bộ các dòng đều mang giờ `07:00:00`. Grain của dự án là "1 mã × 1 ngày"
nên phần giờ này là nhiễu — nếu đưa nguyên vào khoá chính sẽ gây rủi ro
trùng lặp / join sai.

→ Bước transform (Ngày 8) BẮT BUỘC chuẩn hoá `time` về date thuần.

### Kiểm tra ngày có bị lệch múi giờ không

Con số `07:00:00` gợi ý có chuyển đổi múi giờ. Cần chắc `2026-01-05 07:00:00`
đúng là phiên ngày 05/01, không phải phiên 04/01 bị đẩy lên.

Cách kiểm chứng: lấy cùng khoảng ngày từ cả hai nguồn, so `close` theo ngày lịch.

<chạy đoạn code so sánh bên dưới rồi điền kết quả vào đây>

Kết luận: <khớp / lệch 1 ngày>

### Các phát hiện khác từ dữ liệu kbs (2026-01-01 → 2026-09-03, mã VNM)

- Tổng 163 dòng, 6 cột: `time, open, high, low, close, volume`
- **Phiên hôm nay (2026-09-03) có vẻ CHƯA đóng cửa**: volume 1.335.700 trong khi
  các phiên khác 2–6 triệu. → Không nên nạp phiên chưa hoàn tất vào warehouse
  như dữ liệu chính thức. Quyết định: <cắt bỏ ngày hôm nay / cho ghi đè ở lần chạy sau>
- **Thiếu ngày 31/08, 01/09, 02/09**: 02/09 là Quốc khánh, sàn nghỉ lễ dài.
  → Đây là thiếu HỢP LỆ, không phải mất dữ liệu.
  → Validator (Ngày 9) phải phân biệt "sàn nghỉ" với "pipeline lỗi"
  → Cần bảng `dim_date` có cờ ngày lễ (Phase B)

### Ghi chú API

Lớp `Vnstock()` đã bị ngừng hỗ trợ từ 31/08/2025. Dùng `Quote` thay thế:
`from vnstock import Quote` (hoặc `from vnstock.api.quote import Quote`).
